crates/telio-nurse/src/rtt/ping.rs

```diff
@@ -36,39 +50,100 @@ impl Ping {
+                .unwrap_or(0u64);
             let _ = node.rtt_histogram.increment(u64_avg);
+            let _ = node
+                .rtt_loss_histogram
+                .increment((100 * results_v4.unsuccessful_pings / self.no_of_tries) as u64);
+        }
+
+        if let Some(results_v6) = dpr.v6 {
+            let u64_avg = results_v6
+                .avg_rtt
+                .map_or(Duration::from_millis(0), |a| a)
+                .as_millis()
+                .try_into()
+                .unwrap_or(0u64);
+            let _ = node.rtt6_histogram.increment(u64_avg);
+            let _ = node
+                .rtt6_loss_histogram
+                .increment((100 * results_v6.unsuccessful_pings / self.no_of_tries) as u64);
         }
     }
 
-    async fn perform_average_rtt(&self, node: IpAddr) -> Option<Duration> {
-        let client = match node {
-            IpAddr::V4(_) => self.client_v4.clone(),
-            IpAddr::V6(_) => self.client_v6.clone(),
+    async fn perform_average_rtt(&self, target: &DualTarget) -> DualPingResults {
+        let mut dpresults = DualPingResults { v4: None, v6: None };
+
+        let ping_action = |client: Arc<Client>, host| async move {
+            let mut results = PingResults {
+                successful_pings: 0,
+                unsuccessful_pings: 0,
+                avg_rtt: None,
+            };
+            let mut pinger = client.pinger(host, PingIdentifier(rand::random())).await;
+            let payload = [0; 56];
+            pinger.timeout(Self::PING_TIMEOUT);
+
+            let mut sum = Duration::default();
+
+            for i in 0..self.no_of_tries {
+                if let Ok((_, duration)) = pinger
+                    .ping(PingSequence(i.try_into().unwrap_or(0)), &payload)
+                    .await
+                {
+                    sum = sum.saturating_add(duration);
+                    results.successful_pings += 1;
+                } else {
+                    results.unsuccessful_pings += 1;
+                }
+            }
+
+            results.avg_rtt = sum.checked_div(results.successful_pings);
+            results
         };
 
-        let mut pinger = client.pinger(node, PingIdentifier(rand::random())).await;
-        let payload = [0; 56];
-        pinger.timeout(Self::PING_TIMEOUT);
-
-        let mut sum = Duration::default();
-        let mut successful_pings = 0;
-
-        for i in 0..self.no_of_tries {
-            if let Ok((_, duration)) = pinger
-                .ping(PingSequence(i.try_into().unwrap_or(0)), &payload)
-                .await
-            {
-                sum = sum.saturating_add(duration);
-                successful_pings += 1;
-            } else {
-                // TODO: Log error
+        match target.get_targets() {
+            Ok(t) => {
+                match t.0 {
+                    IpAddr::V4(_) => {
+                        dpresults.v4 = Some(ping_action(self.client_v4.clone(), t.0).await);
```

Comment:
isn't .0 is always ipv4 and .1 is always ipv6?

Reply:
Nope, on IPv6 stack, the primary address is IPv6:

```
if let (Some(ip4), Some(ip6)) = (self.target.0, self.target.1) {
    // IPv6 target is preffered, because we can be sure, that address is unique
    Ok((IpAddr::V6(ip6), Some(IpAddr::V4(ip4))))
} else if let (Some(ip4), None) = (self.target.0, self.target.1) {
    Ok((IpAddr::V4(ip4), None))
} else if let (None, Some(ip6)) = (self.target.0, self.target.1) {
    Ok((IpAddr::V6(ip6), None))
} else {
    Err(DualTargetError::NoTarget)
}
```